internal/controller/controllers/crd_utils.go

```diff
@@ -50,15 +62,35 @@ func (u *CRDUtils) CreateAgentCR(ctx context.Context, log logrus.FieldLogger, ho
 			},
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      hostId,
-				Namespace: clusterNamespace,
+				Namespace: infraEnv.Namespace,
 			},
 		}
-		log.Infof("Creating Agent CR. Namespace: %s, Cluster: %s, HostID: %s", clusterNamespace, clusterName, hostId)
+		log.Infof("Creating Agent CR. Namespace: %s, Cluster: %s, HostID: %s", infraEnv.Namespace, clusterName, hostId)
 		return u.client.Create(ctx, host)
 	}
 	return err
 }
 
+// Returns the InfraEnv resource associated with the cluster
+func (u *CRDUtils) getInfraEnvByCluster(ctx context.Context, log logrus.FieldLogger, clusterNamespace, clusterName string) (*aiv1beta1.InfraEnv, error) {
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	deploymentNamespacedName := types.NamespacedName{
+		Name:      clusterName,
+		Namespace: clusterNamespace,
+	}
+	err := u.client.Get(ctx, deploymentNamespacedName, clusterDeployment)
```

Comment:
The only think that is required in order to get infra env is the cluster deployment name and namespace so i don't think that it's required to get the cluster deployment, maybe instead you should fix `getInfraEnvByClusterDeployment` and don't call for cluster deployment get. 
In addition i think that there is a problem in `getInfraEnvByClusterDeployment` it looks like it ocmpare only names and not the namespace. you could have multiple cluster deployments with the same name and a different namespace.

Reply:
Good point, simplified by changing getInfraEnvByClusterDeployment.